src/lib.rs

```diff
@@ -475,496 +528,62 @@ unsafe impl<'a> Slab for BorrowedRawAllocation<'a> {
     }
 }
 
-/// Given pointer and offset, returns a new offset aligned to `align`.
-///
-/// `align` *must* be a power of two and >= 1 or else the result is meaningless.
-fn align_offset_up_to(ptr: usize, offset: usize, align: usize) -> Option<usize> {
-    let offsetted_ptr = ptr.checked_add(offset)?;
-    let aligned_ptr = offsetted_ptr.checked_add(align - 1)? & !(align - 1);
-    // don't need to check since we know aligned_ptr is >= ptr at this point
-    Some(aligned_ptr - ptr)
+/// Computed offsets necessary for a copy or read operation with some layout. Should only be
+/// created by [`compute_offsets`]
+#[derive(Debug, Copy, Clone)]
+pub(crate) struct ComputedOffsets {
+    start: usize,
+    end: usize,
+    end_padded: usize,
 }
 
-/// Compute and validate offsets for a copy operation with the given parameters.
-fn compute_offsets<S: Slab>(
-    dst: &S,
+/// Compute and validate offsets for a copy or read operation with the given parameters.
+pub(crate) fn compute_and_validate_offsets<S: Slab>(
+    slab: &S,
     start_offset: usize,
     t_layout: Layout,
     min_alignment: usize,
     require_exact_start_offset: bool,
-) -> Result<CopyRecord, CopyError> {
-    let copy_layout = t_layout.align_to(min_alignment.next_power_of_two())?;
-
-    let copy_start_offset =
-        align_offset_up_to(dst.base_ptr() as usize, start_offset, copy_layout.align())
-            .ok_or(CopyError::InvalidLayout)?;
-    if require_exact_start_offset && start_offset != copy_start_offset {
-        return Err(CopyError::RequestedOffsetUnaligned);
+) -> Result<ComputedOffsets, Error> {
+    let layout = t_layout.align_to(min_alignment.next_power_of_two())?;
```

Comment:
Fwiw I feel like the caller could have already re-aligned their `Layout` argument if they needed?

Reply:
yeah but the point of this function is to unify the layout logic as much as possible so that it can be reviewed/verified once and used in many places